core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -64,672 +83,1650 @@
+                        future.set(null);
+                    }
+                });
+            }
+
+            return future;
+        }
+
+        private PipelinedDistributedStagesScheduler(
+                DistributedStagesSchedulerStateMachine stateMachine,
+                QueryStateMachine queryStateMachine,
+                SplitSchedulerStats schedulerStats,
+                StageManager stageManager,
+                ExecutionSchedule executionSchedule,
+                Map<StageId, StageScheduler> stageSchedulers,
+                Map<StageId, PipelinedStageExecution> stageExecutions,
+                DynamicFilterService dynamicFilterService)
+        {
+            this.stateMachine = requireNonNull(stateMachine, "stateMachine is null");
+            this.queryStateMachine = requireNonNull(queryStateMachine, "queryStateMachine is null");
+            this.schedulerStats = requireNonNull(schedulerStats, "schedulerStats is null");
+            this.stageManager = requireNonNull(stageManager, "stageManager is null");
+            this.executionSchedule = requireNonNull(executionSchedule, "executionSchedule is null");
+            this.stageSchedulers = ImmutableMap.copyOf(requireNonNull(stageSchedulers, "stageSchedulers is null"));
+            this.stageExecutions = ImmutableMap.copyOf(requireNonNull(stageExecutions, "stageExecutions is null"));
+            this.dynamicFilterService = requireNonNull(dynamicFilterService, "dynamicFilterService is null");
+        }
+
+        private void initialize()
+        {
+            for (PipelinedStageExecution stageExecution : stageExecutions.values()) {
+                List<PipelinedStageExecution> childStageExecutions = stageManager.getChildren(stageExecution.getStageId()).stream()
+                        .map(stage -> requireNonNull(stageExecutions.get(stage.getStageId()), () -> "stage execution not found for stage: " + stage))
+                        .collect(toImmutableList());
+                if (!childStageExecutions.isEmpty()) {
+                    stageExecution.addStateChangeListener(newState -> {
+                        if (newState == FLUSHING || newState.isDone()) {
+                            childStageExecutions.forEach(PipelinedStageExecution::cancel);
+                        }
+                    });
+                }
+            }
+
+            Set<StageId> finishedStages = newConcurrentHashSet();
+            for (PipelinedStageExecution stageExecution : stageExecutions.values()) {
+                stageExecution.addStateChangeListener(state -> {
+                    if (stateMachine.getState().isDone()) {
+                        return;
+                    }
+                    // TODO: support dynamic filter for failure retries
+                    if (!state.canScheduleMoreTasks()) {
+                        dynamicFilterService.stageCannotScheduleMoreTasks(stageExecution.getStageId(), stageExecution.getAllTasks().size());
```

Comment:
Can we re-use result of `stageExecution.getAllTasks().size()` for the `!stageExecution.getAllTasks().isEmpty()` check on next line as well ?